File: patterns/02-reliability-patterns/04-bulkhead/go/bulkhead.go
package bulkhead

type PatternError struct{ code string }

func (e *PatternError) Error() string { return e.code }
func (e *PatternError) Code() string  { return e.code }
func Evaluate(input map[string]any) (any, error) {
	capacities := map[string]int{}
	used := map[string]int{}
	for key, value := range input["capacities"].(map[string]any) {
		capacities[key] = int(value.(float64))
		used[key] = 0
	}
	decisions := []string{}
	raw, _ := input["actions"].([]any)
	for _, item := range raw {
		action := item.(map[string]any)
		pool := action["pool"].(string)
		capacity, ok := capacities[pool]
		if !ok {
			return nil, &PatternError{code: "unknown_pool"}
		}
		switch action["op"] {
		case "acquire":
			if used[pool] >= capacity {
				decisions = append(decisions, "rejected:"+pool)
			} else {
				used[pool]++
				decisions = append(decisions, "accepted:"+pool)
			}
		case "release":
			if used[pool] <= 0 {
				return nil, &PatternError{code: "over_release"}
			}
			used[pool]--
			decisions = append(decisions, "released:"+pool)
		default:
			return nil, &PatternError{code: "unsupported_action"}
		}
	}
	out := map[string]any{}
	for key, value := range used {
		out[key] = value
	}
	return map[string]any{"decisions": decisions, "used": out}, nil
}

File: patterns/02-reliability-patterns/04-bulkhead/python/bulkhead.py
class PatternError(Exception):
    def __init__(self,code):super().__init__(code);self.code=code
def evaluate(input_data):
    capacities=dict(input_data.get("capacities",{}));used={key:0 for key in capacities};decisions=[]
    for action in input_data.get("actions",[]):
        pool=action["pool"]
        if pool not in capacities:raise PatternError("unknown_pool")
        if action["op"]=="acquire":
            if used[pool]>=capacities[pool]:decisions.append(f"rejected:{pool}")
            else:used[pool]+=1;decisions.append(f"accepted:{pool}")
        elif action["op"]=="release":
            if used[pool]<=0:raise PatternError("over_release")
            used[pool]-=1;decisions.append(f"released:{pool}")
        else:raise PatternError("unsupported_action")
    return{"decisions":decisions,"used":dict(sorted(used.items()))}
